mantidproject/mantid
# Fix OpenGL segmentation fault on Linux

When drawing some instruments in full 3d (with OpenGL), putting all the draw commands for the instrument in one big list seems to cause a segmentation fault on certain versions of Linux, including the version used on IDAaaS. This changes `InstrumentRenderer` to draw each component of the instrument using a separate list, when the `MantidOptions.InstrumentView.MesaBugWorkaround` option in the user properties file is set to `On`.

The original bug in Mesa was fixed [here](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/27556), but that version has not made it onto IDAaaS yet.

#### Summary of work

- Draw each instrument component in a separate display list when the new option is enabled.
- If option is disabled then behaviour should be the same.
- Add a method that will give you the OpenGL version string (someone will thank me in the future).
- Split the picking and non-picking display lists into named member variables instead of using two indices.

Fixes #36220 

### To test:
See #36220, or what caused the original segmentation fault for me was to view full 3D for MAR11001 (on Rocky 8 IDAaaS and Ubuntu). Other instruments and files will need checking as well.

Updated conda packages have bee uploaded to thomashampson/label/mesa_workaround:
`mamba install -c thomashampson/label/opengl_segfault_fix mantidworkbench`
Install this on IDAaaS and check that it doesn't crash.







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
